src/main/java/org/radarcns/android/MainActivity.java

```diff
@@ -101,141 +64,74 @@
-     * for this.
-     *
-     * @return configured RadarConfiguration
-     */
-    protected abstract RadarConfiguration createConfiguration();
+
 
     protected abstract Class<? extends LoginActivity> loginActivity();
 
     @Override
-    protected void onCreate(Bundle savedInstanceState) {
+    protected final void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
-        radarConfiguration = createConfiguration();
-
         authState = AppAuthState.Builder.from(this).build();
         if (!authState.isValid()) {
             startLogin(false);
             return;
         }
-        radarConfiguration.put(RadarConfiguration.PROJECT_ID_KEY, authState.getProjectId());
-        radarConfiguration.put(RadarConfiguration.USER_ID_KEY, authState.getUserId());
 
-        onConfigChanged();
-        logger.info("RADAR configuration at create: {}", radarConfiguration);
+        create();
+    }
 
-        try {
-            mConnections = DeviceServiceProvider.loadProviders(this, radarConfiguration);
-            // TODO: check what sources are available in the authState (if any)
-            // if any sources are available:
-            //   - only start up providers that DeviceServiceProvider#matches one of the sources
-            //   - set that source in the DeviceServiceProvider.
-            for (DeviceServiceProvider provider : mConnections) {
-                DeviceServiceConnection connection = provider.getConnection();
-                mTotalRecordsSent.put(connection, new TimedInt());
-                deviceFilters.put(connection, Collections.<String>emptySet());
-            }
-        } catch (IllegalArgumentException ex) {
-            logger.error("Cannot instantiate device provider, waiting to fetch to complete", ex);
-            mConnections = Collections.emptyList();
+    @CallSuper
+    protected void create() {
+        RadarConfiguration radarConfiguration = RadarConfiguration.getInstance();
+
+        if (authState.getProjectId() != null) {
+            radarConfiguration.put(RadarConfiguration.PROJECT_ID_KEY, authState.getProjectId());
         }
+        radarConfiguration.put(RadarConfiguration.USER_ID_KEY, authState.getUserId());
+        logger.info("RADAR configuration at create: {}", radarConfiguration);
+        onConfigChanged();
 
-        radarConfiguration.onFetchComplete(this, new OnCompleteListener<Void>() {
+        configurationBroadcastReceiver = new BroadcastReceiver() {
             @Override
-            public void onComplete(@NonNull Task<Void> task) {
-                if (task.isSuccessful()) {
-                    // Once the config is successfully fetched it must be
-                    // activated before newly fetched values are returned.
-                    radarConfiguration.activateFetched();
-                    for (DeviceServiceProvider provider : mConnections) {
-                        provider.updateConfiguration();
```

Comment:
How are the device service configurations now updated and activated if the Firebase configuration changed? It would be good to solve that in commons, rather than in the individual pRMT apps

Reply:
That logic is now in RadarService. See RadarService.onCreate and RadarService.configure

Reply:
Ok, I missed that, good stuff!